From c3e4b9a42a94bbfd9230171b0d068b0b27756d27 Mon Sep 17 00:00:00 2001
From: Mihai Dontu <mdontu@bitdefender.com>
Date: Mon, 29 Jun 2015 14:49:20 +0300
Subject: [PATCH 23/26] added support for skipping instructions
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

There are situations when we need to skip over an instruction in order
to prevent it from accessing the memory in any way (read/fetch or
write).

Based on Razvan Cojocaru's patch:

 xen: Disable HV-based instruction length computing

Signed-off-by: Mihai Don»õu <mdontu@bitdefender.com>
---
 xen/arch/x86/mm/p2m.c         | 21 ++++++++++++---------
 xen/include/public/vm_event.h |  5 +++++
 2 files changed, 17 insertions(+), 9 deletions(-)

diff --git a/xen/arch/x86/mm/p2m.c b/xen/arch/x86/mm/p2m.c
index 46e3ed1..87b71a0 100644
--- a/xen/arch/x86/mm/p2m.c
+++ b/xen/arch/x86/mm/p2m.c
@@ -1430,6 +1430,8 @@ void p2m_mem_access_emulate_check(struct vcpu *v,
         if ( (rsp->flags & VM_EVENT_FLAG_SET_EMUL_READ_DATA) &&
              v->arch.vm_event.emul_read_data )
             *v->arch.vm_event.emul_read_data = rsp->data.emul_read_data;
+        else if ( rsp->flags & VM_EVENT_FLAG_SKIP_INSTRUCTION )
+            v->arch.vm_event.eip = rsp->data.regs.x86.rsp;
     }
 }
 
@@ -1516,17 +1518,18 @@ bool_t p2m_mem_access_check(paddr_t gpa, unsigned long gla,
 
     if ( v->arch.vm_event.emulate_flags )
     {
-        enum emul_kind kind = EMUL_KIND_NORMAL;
-
-        if ( v->arch.vm_event.emulate_flags &
-             VM_EVENT_FLAG_SET_EMUL_READ_DATA )
-            kind = EMUL_KIND_SET_CONTEXT;
+        if ( v->arch.vm_event.emulate_flags & VM_EVENT_FLAG_SKIP_INSTRUCTION )
+            guest_cpu_user_regs()->eip = v->arch.vm_event.eip;
+        else if ( v->arch.vm_event.emulate_flags & VM_EVENT_FLAG_SET_EMUL_READ_DATA )
+            hvm_mem_access_emulate_one(EMUL_KIND_SET_CONTEXT, TRAP_invalid_op,
+                                       HVM_DELIVER_NO_ERROR_CODE);
         else if ( v->arch.vm_event.emulate_flags &
                   VM_EVENT_FLAG_EMULATE_NOWRITE )
-            kind = EMUL_KIND_NOWRITE;
-
-        hvm_mem_access_emulate_one(kind, TRAP_invalid_op,
-                                   HVM_DELIVER_NO_ERROR_CODE);
+            hvm_mem_access_emulate_one(EMUL_KIND_NOWRITE, TRAP_invalid_op,
+                                       HVM_DELIVER_NO_ERROR_CODE);
+        else
+            hvm_mem_access_emulate_one(EMUL_KIND_NORMAL, TRAP_invalid_op,
+                                       HVM_DELIVER_NO_ERROR_CODE);
 
         v->arch.vm_event.emulate_flags = 0;
         return 1;
diff --git a/xen/include/public/vm_event.h b/xen/include/public/vm_event.h
index fbc76b2..4091b4e 100644
--- a/xen/include/public/vm_event.h
+++ b/xen/include/public/vm_event.h
@@ -81,6 +81,11 @@
 #define VM_EVENT_FLAG_DENY               (1 << 6)
 
 /*
+ * Skip the current instruction, avoiding any potential read/fetch side-effects
+ */
+#define VM_EVENT_FLAG_SKIP_INSTRUCTION   (1 << 7)
+
+/*
  * Reasons for the vm event request
  */
 
-- 
2.5.0

